datafusion/optimizer/src/optimizer.rs

```diff
@@ -299,44 +379,65 @@ impl Optimizer {
             log_plan(&format!("Optimizer input (pass {i})"), &new_plan);
 
             for rule in &self.rules {
-                let result =
-                    self.optimize_recursively(rule, &new_plan, config)
-                        .and_then(|plan| {
-                            if let Some(plan) = &plan {
-                                assert_schema_is_the_same(rule.name(), plan, &new_plan)?;
-                            }
-                            Ok(plan)
-                        });
-                match result {
-                    Ok(Some(plan)) => {
+                // If we need to skip failed rules, must copy plan before attempting to rewrite
+                // as rewriting is destructive
+                let prev_plan = options
```

Comment:
Now we only copy the (entire!) plan if we need to be able to rewrite it. Otherwise no copy is done